src/charm.py

```diff
@@ -351,6 +349,16 @@ def _single_mongo_replica(self, ip_address: str) -> MongoDB:
         unit_config["calling_unit_ip"] = ip_address
         return MongoDB(unit_config)
 
+    def _unit_ip(self, unit: ops.model.Unit) -> str:
+        """Returns the ip address of a given unit."""
+        if unit == self.unit:
+            return str(self.model.get_binding(PEER).network.bind_address)
+
+        if unit in self._peers.data:
+            return str(self._peers.data[unit].get("private-address"))
+
+        return None
+
```

Comment:
I quite like this encapsulation.
I would suggest changing the exit-points to be an:
```
if:
elif:
else: error
```

just in case returning 'None' wrecks things or upsets linters.